### DISCUSSION TITLE
My code uses lots of `!`

### DISCUSSION THREAD

**Discussion by Zambrella** [serverpod/serverpod#1805] [Q&A] [answered]
I have quite a lot of relations going on (sometimes nested up to 4 objects deep) and find myself having to "cast away nullability" a lot (especially as I have them [defined on both sides](https://docs.serverpod.dev/concepts/database/relations/one-to-one#independent-relations-defined-on-both-sides) in both server and client code. This isn't too bad in the moment but when it comes to maintaining and building upon the code I have to go back and confirm to myself whether I did include the relation or not.

What I'm asking is if this is a sign I'm doing something wrong or just a limitation of the ORM?

I suppose one solution is just robust tests and checking for null with if statements and/or asserts?

Thanks in advance.

**Comment by jason-downs** [comment]
Yes, I am experiencing this too since the update to 1.2.x, with the welcome addition of more database relation functionality, it has also meant I'm doing an awful lot of ! in my code which I can live with for now but is not ideal, especially since it kind of seems unnecessary since these aren't meant to be nullable fields and is only becuase it is required to be nullable to have a db relation.

**Comment by SandPod** [comment, marked as answer]
Hey @Zambrella!

Really happy to hear you are exploring and utilizing the new relation system.

This is a really interesting question and there are probably good ideas for how the UX can improve here.

### Background
The simple motivation behind always having the relation field as nullable is that we want it to be the developers choice to include the related data with the query. The nullability of the field enables the framework to not require related data in object deserialization.

### Workarounds
#### Nullable relation
If you have an optional relation, where the DB actually could return null, you should probably keep null checking like you are doing today. To alleviate the pain the scary `!` I would recommend favoring `?` and then doing an explicit null check at the use place.

#### DTO
If you would like to resolve all your null checks in a single place and then pass down an object that does not have nullable fields (for example sending it to the client), you could define a new Model that is not stored in the database. The fields of the model could hold the models you would like to pass on but then be none nullable. Then once you have fetched the data from the database, you could do your null checks and then instantiate an object of that type instead and pass it allong.

### Future
It would be really cool to be able provide non nullable fields if we know that the information is guaranteed to be there. We have been discussing a bunch of potential solutions for this but I would be nice to gather your suggestions as well. Here is a small list of ideas that have been floating around:

- Views: Have the possibility to define views that combine columns from different tables and are also serialized to models.
- Predefined queries: Have the possibility to create predefined queries that are then represented as a Data object that could have non nullable fields if the include is non nullable.

**Reply by Zambrella** [reply]
Thanks for getting back so quickly!

I really like the idea of "views" because, in my use-case at least, there are a handful of core tables that store the most important information but the client/UI needs that data in multiple different ways. It sounds like "views" would allow me to keep my (hopefully) well structured database schema but give me the flexibility to mix and match the data into models that make sense for the client.

My above scenario does seem like DTOs would be the way to go but it feels like I'm then losing the benefit of using an all-Dart ecosystem/server-framework.

_Side-note: I'm quite new to relational databases and server development in general._

**Reply by SandPod** [reply]
> My above scenario does seem like DTOs would be the way to go but it feels like I'm then losing the benefit of using an all-Dart ecosystem/server-framework.

If you define the DTO as a Serverpod model without a table, you should still benefit from the model generation and serialization that the framework provides as you can use it directly as an input parameter or return type from any of your endpoints.

> Side-note: I'm quite new to relational databases and server development in general.

That is awesome! One of the goals of the framework is to make server side development as intuitive and accessible as possible but still include all power features.
